File: cpp/983_MinimumCostForTickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int last = days[days.size()-1];
        vector<int> dp(last+1, 0);
        unordered_set<int> all_days;
        
        for (int d : days)
            all_days.insert(d);
        
        for (int i = 1; i < last+1; ++i) {
            if (all_days.find(i) == all_days.end()) {
                dp[i] = dp[i-1];
            } else {
                int one = dp[i-1] + costs[0];
                int seven = dp[max(0, i-7)] + costs[1];
                int thirty = dp[max(0, i-30)] + costs[2];
                dp[i] = min(one, seven, thirty);
            }
        }
        return dp[dp.size()-1];
    }
    
    int min(const int& a, const int& b, const int& c) {
        return a <= b  && a <= c ? a : b < c ? b : c;
    }
};

File: python/983_MinimumCostForTickets.py
from collections import Counter
class Solution:
    def mincostTickets(self, days: List[int], cost: List[int]) -> int:
        last = days[-1]
        dp = [0 for _ in range(last+1)]
        travel_days = Counter(days)
        
        for i in range(last+1):
            if i not in travel_days:
                dp[i] = dp[i-1]
            else:
                one, seven, thirty = dp[max(0, i-1)] + cost[0], dp[max(0, i-7)] + cost[1], dp[max(0, i-30)] + cost[2]
                dp[i] = min(one, seven, thirty)
        
        return dp[-1]
                
            
